// wechaty/dist/src/user/mini-program.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../config");
class MiniProgram {
    /*
     * @hideconstructor
     */
    constructor(payload) {
        this.payload = payload;
        config_1.log.verbose('MiniProgram', 'constructor()');
    }
    /**
     *
     * Create
     *
     */
    static create() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('MiniProgram', 'create()');
            // TODO: get appid and username from wechat
            const payload = {
                appid: 'todo',
                description: 'todo',
                pagePath: 'todo',
                thumbKey: 'todo',
                thumbUrl: 'todo',
                title: 'todo',
                username: 'todo',
            };
            return new MiniProgram(payload);
        });
    }
    appid() {
        return this.payload.appid;
    }
    title() {
        return this.payload.title;
    }
    pagePath() {
        return this.payload.pagePath;
    }
    username() {
        return this.payload.username;
    }
    description() {
        return this.payload.description;
    }
    thumbUrl() {
        return this.payload.thumbUrl;
    }
    thumbKey() {
        return this.payload.thumbKey;
    }
}
exports.MiniProgram = MiniProgram;
