// @chatie/grpc/dist/generated/wechaty/puppet_grpc_pb.js
// GENERATED CODE -- DO NOT EDIT!

// Original file comments:
// *
//
// Wechaty Puppet gRPC Protocol Buffers
//  https://github.com/chatie/grpc/
//  Huan LI <zixia@zixia.net>
//  Apr 2018
//  License: Apache-2.0
//
// Google Protocol Buffers
//  Style Guide - https://developers.google.com/protocol-buffers/docs/style
//
'use strict';
var grpc = require('grpc');
var wechaty_puppet_base_pb = require('../wechaty/puppet/base_pb.js');
var wechaty_puppet_contact_pb = require('../wechaty/puppet/contact_pb.js');
var wechaty_puppet_event_pb = require('../wechaty/puppet/event_pb.js');
var wechaty_puppet_friendship_pb = require('../wechaty/puppet/friendship_pb.js');
var wechaty_puppet_message_pb = require('../wechaty/puppet/message_pb.js');
var wechaty_puppet_room_pb = require('../wechaty/puppet/room_pb.js');
var wechaty_puppet_room_invitation_pb = require('../wechaty/puppet/room_invitation_pb.js');
var wechaty_puppet_room_member_pb = require('../wechaty/puppet/room_member_pb.js');
var wechaty_puppet_tag_pb = require('../wechaty/puppet/tag_pb.js');

function serialize_wechaty_puppet_base_DingRequest(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.DingRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.base.DingRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_DingRequest(buffer_arg) {
  return wechaty_puppet_base_pb.DingRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_DingResponse(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.DingResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.base.DingResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_DingResponse(buffer_arg) {
  return wechaty_puppet_base_pb.DingResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_LogoutRequest(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.LogoutRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.base.LogoutRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_LogoutRequest(buffer_arg) {
  return wechaty_puppet_base_pb.LogoutRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_LogoutResponse(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.LogoutResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.base.LogoutResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_LogoutResponse(buffer_arg) {
  return wechaty_puppet_base_pb.LogoutResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_StartRequest(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.StartRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.base.StartRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_StartRequest(buffer_arg) {
  return wechaty_puppet_base_pb.StartRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_StartResponse(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.StartResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.base.StartResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_StartResponse(buffer_arg) {
  return wechaty_puppet_base_pb.StartResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_StopRequest(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.StopRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.base.StopRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_StopRequest(buffer_arg) {
  return wechaty_puppet_base_pb.StopRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_StopResponse(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.StopResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.base.StopResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_StopResponse(buffer_arg) {
  return wechaty_puppet_base_pb.StopResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_VersionRequest(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.VersionRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.base.VersionRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_VersionRequest(buffer_arg) {
  return wechaty_puppet_base_pb.VersionRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_base_VersionResponse(arg) {
  if (!(arg instanceof wechaty_puppet_base_pb.VersionResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.base.VersionResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_base_VersionResponse(buffer_arg) {
  return wechaty_puppet_base_pb.VersionResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactAliasRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactAliasRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactAliasRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactAliasRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactAliasRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactAliasResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactAliasResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactAliasResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactAliasResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactAliasResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactAvatarRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactAvatarRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactAvatarRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactAvatarRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactAvatarRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactAvatarResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactAvatarResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactAvatarResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactAvatarResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactAvatarResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactListRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactListRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactListRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactListRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactListRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactListResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactListResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactListResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactListResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactListResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactPayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactPayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactPayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactPayloadRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactPayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactPayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactPayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactPayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactPayloadResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactPayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfNameRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfNameRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfNameRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfNameRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfNameRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfNameResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfNameResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfNameResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfNameResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfNameResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfQRCodeRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfQRCodeRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfQRCodeRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfQRCodeRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfQRCodeRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfQRCodeResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfQRCodeResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfQRCodeResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfQRCodeResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfQRCodeResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfSignatureRequest(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfSignatureRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfSignatureRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfSignatureRequest(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfSignatureRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_contact_ContactSelfSignatureResponse(arg) {
  if (!(arg instanceof wechaty_puppet_contact_pb.ContactSelfSignatureResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.contact.ContactSelfSignatureResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_contact_ContactSelfSignatureResponse(buffer_arg) {
  return wechaty_puppet_contact_pb.ContactSelfSignatureResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_event_EventRequest(arg) {
  if (!(arg instanceof wechaty_puppet_event_pb.EventRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.event.EventRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_event_EventRequest(buffer_arg) {
  return wechaty_puppet_event_pb.EventRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_event_EventResponse(arg) {
  if (!(arg instanceof wechaty_puppet_event_pb.EventResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.event.EventResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_event_EventResponse(buffer_arg) {
  return wechaty_puppet_event_pb.EventResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipAcceptRequest(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipAcceptRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipAcceptRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipAcceptRequest(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipAcceptRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipAcceptResponse(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipAcceptResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipAcceptResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipAcceptResponse(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipAcceptResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipAddRequest(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipAddRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipAddRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipAddRequest(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipAddRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipAddResponse(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipAddResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipAddResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipAddResponse(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipAddResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipPayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipPayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipPayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipPayloadRequest(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipPayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipPayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipPayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipPayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipPayloadResponse(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipPayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipSearchPhoneRequest(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipSearchPhoneRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipSearchPhoneRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipSearchPhoneRequest(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipSearchPhoneRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipSearchPhoneResponse(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipSearchPhoneResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipSearchPhoneResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipSearchPhoneResponse(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipSearchPhoneResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipSearchWeixinRequest(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipSearchWeixinRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipSearchWeixinRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipSearchWeixinRequest(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipSearchWeixinRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_friendship_FriendshipSearchWeixinResponse(arg) {
  if (!(arg instanceof wechaty_puppet_friendship_pb.FriendshipSearchWeixinResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.friendship.FriendshipSearchWeixinResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_friendship_FriendshipSearchWeixinResponse(buffer_arg) {
  return wechaty_puppet_friendship_pb.FriendshipSearchWeixinResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageContactRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageContactRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageContactRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageContactRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageContactRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageContactResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageContactResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageContactResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageContactResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageContactResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageFileRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageFileRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageFileRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageFileRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageFileRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageFileResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageFileResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageFileResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageFileResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageFileResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageMiniProgramRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageMiniProgramRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageMiniProgramRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageMiniProgramRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageMiniProgramRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageMiniProgramResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageMiniProgramResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageMiniProgramResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageMiniProgramResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageMiniProgramResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessagePayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessagePayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessagePayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessagePayloadRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessagePayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessagePayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessagePayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessagePayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessagePayloadResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessagePayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageRecallRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageRecallRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageRecallRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageRecallRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageRecallRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageRecallResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageRecallResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageRecallResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageRecallResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageRecallResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendContactRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendContactRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendContactRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendContactRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendContactRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendContactResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendContactResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendContactResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendContactResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendContactResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendFileRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendFileRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendFileRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendFileRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendFileRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendFileResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendFileResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendFileResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendFileResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendFileResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendMiniProgramRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendMiniProgramRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendMiniProgramRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendMiniProgramRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendMiniProgramRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendMiniProgramResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendMiniProgramResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendMiniProgramResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendMiniProgramResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendMiniProgramResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendTextRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendTextRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendTextRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendTextRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendTextRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendTextResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendTextResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendTextResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendTextResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendTextResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendUrlRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendUrlRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendUrlRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendUrlRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendUrlRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageSendUrlResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageSendUrlResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageSendUrlResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageSendUrlResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageSendUrlResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageUrlRequest(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageUrlRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageUrlRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageUrlRequest(buffer_arg) {
  return wechaty_puppet_message_pb.MessageUrlRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_message_MessageUrlResponse(arg) {
  if (!(arg instanceof wechaty_puppet_message_pb.MessageUrlResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.message.MessageUrlResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_message_MessageUrlResponse(buffer_arg) {
  return wechaty_puppet_message_pb.MessageUrlResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAddRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAddRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAddRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAddRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAddRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAddResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAddResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAddResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAddResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAddResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAnnounceRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAnnounceRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAnnounceRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAnnounceRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAnnounceRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAnnounceResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAnnounceResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAnnounceResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAnnounceResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAnnounceResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAvatarRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAvatarRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAvatarRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAvatarRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAvatarRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomAvatarResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomAvatarResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomAvatarResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomAvatarResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomAvatarResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomCreateRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomCreateRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomCreateRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomCreateRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomCreateRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomCreateResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomCreateResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomCreateResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomCreateResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomCreateResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomDelRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomDelRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomDelRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomDelRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomDelRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomDelResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomDelResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomDelResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomDelResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomDelResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomListRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomListRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomListRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomListRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomListRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomListResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomListResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomListResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomListResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomListResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomPayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomPayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomPayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomPayloadRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomPayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomPayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomPayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomPayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomPayloadResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomPayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomQRCodeRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomQRCodeRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomQRCodeRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomQRCodeRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomQRCodeRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomQRCodeResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomQRCodeResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomQRCodeResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomQRCodeResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomQRCodeResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomQuitRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomQuitRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomQuitRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomQuitRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomQuitRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomQuitResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomQuitResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomQuitResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomQuitResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomQuitResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomTopicRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomTopicRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomTopicRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomTopicRequest(buffer_arg) {
  return wechaty_puppet_room_pb.RoomTopicRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_RoomTopicResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_pb.RoomTopicResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room.RoomTopicResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_RoomTopicResponse(buffer_arg) {
  return wechaty_puppet_room_pb.RoomTopicResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_invitation_RoomInvitationAcceptRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_invitation_pb.RoomInvitationAcceptRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room_invitation.RoomInvitationAcceptRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_invitation_RoomInvitationAcceptRequest(buffer_arg) {
  return wechaty_puppet_room_invitation_pb.RoomInvitationAcceptRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_invitation_RoomInvitationAcceptResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_invitation_pb.RoomInvitationAcceptResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room_invitation.RoomInvitationAcceptResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_invitation_RoomInvitationAcceptResponse(buffer_arg) {
  return wechaty_puppet_room_invitation_pb.RoomInvitationAcceptResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_invitation_RoomInvitationPayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_invitation_pb.RoomInvitationPayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room_invitation.RoomInvitationPayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_invitation_RoomInvitationPayloadRequest(buffer_arg) {
  return wechaty_puppet_room_invitation_pb.RoomInvitationPayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_invitation_RoomInvitationPayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_invitation_pb.RoomInvitationPayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room_invitation.RoomInvitationPayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_invitation_RoomInvitationPayloadResponse(buffer_arg) {
  return wechaty_puppet_room_invitation_pb.RoomInvitationPayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_member_RoomMemberListRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_member_pb.RoomMemberListRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room_member.RoomMemberListRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_member_RoomMemberListRequest(buffer_arg) {
  return wechaty_puppet_room_member_pb.RoomMemberListRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_member_RoomMemberListResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_member_pb.RoomMemberListResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room_member.RoomMemberListResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_member_RoomMemberListResponse(buffer_arg) {
  return wechaty_puppet_room_member_pb.RoomMemberListResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_member_RoomMemberPayloadRequest(arg) {
  if (!(arg instanceof wechaty_puppet_room_member_pb.RoomMemberPayloadRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.room_member.RoomMemberPayloadRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_member_RoomMemberPayloadRequest(buffer_arg) {
  return wechaty_puppet_room_member_pb.RoomMemberPayloadRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_room_member_RoomMemberPayloadResponse(arg) {
  if (!(arg instanceof wechaty_puppet_room_member_pb.RoomMemberPayloadResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.room_member.RoomMemberPayloadResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_room_member_RoomMemberPayloadResponse(buffer_arg) {
  return wechaty_puppet_room_member_pb.RoomMemberPayloadResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactAddRequest(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactAddRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactAddRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactAddRequest(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactAddRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactAddResponse(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactAddResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactAddResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactAddResponse(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactAddResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactDeleteRequest(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactDeleteRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactDeleteRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactDeleteRequest(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactDeleteRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactDeleteResponse(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactDeleteResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactDeleteResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactDeleteResponse(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactDeleteResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactListRequest(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactListRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactListRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactListRequest(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactListRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactListResponse(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactListResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactListResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactListResponse(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactListResponse.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactRemoveRequest(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactRemoveRequest)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactRemoveRequest');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactRemoveRequest(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactRemoveRequest.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_wechaty_puppet_tag_TagContactRemoveResponse(arg) {
  if (!(arg instanceof wechaty_puppet_tag_pb.TagContactRemoveResponse)) {
    throw new Error('Expected argument of type wechaty.puppet.tag.TagContactRemoveResponse');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_wechaty_puppet_tag_TagContactRemoveResponse(buffer_arg) {
  return wechaty_puppet_tag_pb.TagContactRemoveResponse.deserializeBinary(new Uint8Array(buffer_arg));
}


var PuppetService = exports.PuppetService = {
  // *
//
// Base
//
start: {
    path: '/wechaty.puppet.Puppet/Start',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_base_pb.StartRequest,
    responseType: wechaty_puppet_base_pb.StartResponse,
    requestSerialize: serialize_wechaty_puppet_base_StartRequest,
    requestDeserialize: deserialize_wechaty_puppet_base_StartRequest,
    responseSerialize: serialize_wechaty_puppet_base_StartResponse,
    responseDeserialize: deserialize_wechaty_puppet_base_StartResponse,
  },
  stop: {
    path: '/wechaty.puppet.Puppet/Stop',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_base_pb.StopRequest,
    responseType: wechaty_puppet_base_pb.StopResponse,
    requestSerialize: serialize_wechaty_puppet_base_StopRequest,
    requestDeserialize: deserialize_wechaty_puppet_base_StopRequest,
    responseSerialize: serialize_wechaty_puppet_base_StopResponse,
    responseDeserialize: deserialize_wechaty_puppet_base_StopResponse,
  },
  logout: {
    path: '/wechaty.puppet.Puppet/Logout',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_base_pb.LogoutRequest,
    responseType: wechaty_puppet_base_pb.LogoutResponse,
    requestSerialize: serialize_wechaty_puppet_base_LogoutRequest,
    requestDeserialize: deserialize_wechaty_puppet_base_LogoutRequest,
    responseSerialize: serialize_wechaty_puppet_base_LogoutResponse,
    responseDeserialize: deserialize_wechaty_puppet_base_LogoutResponse,
  },
  ding: {
    path: '/wechaty.puppet.Puppet/Ding',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_base_pb.DingRequest,
    responseType: wechaty_puppet_base_pb.DingResponse,
    requestSerialize: serialize_wechaty_puppet_base_DingRequest,
    requestDeserialize: deserialize_wechaty_puppet_base_DingRequest,
    responseSerialize: serialize_wechaty_puppet_base_DingResponse,
    responseDeserialize: deserialize_wechaty_puppet_base_DingResponse,
  },
  version: {
    path: '/wechaty.puppet.Puppet/Version',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_base_pb.VersionRequest,
    responseType: wechaty_puppet_base_pb.VersionResponse,
    requestSerialize: serialize_wechaty_puppet_base_VersionRequest,
    requestDeserialize: deserialize_wechaty_puppet_base_VersionRequest,
    responseSerialize: serialize_wechaty_puppet_base_VersionResponse,
    responseDeserialize: deserialize_wechaty_puppet_base_VersionResponse,
  },
  // *
//
// Event - Server Stream
//
event: {
    path: '/wechaty.puppet.Puppet/Event',
    requestStream: false,
    responseStream: true,
    requestType: wechaty_puppet_event_pb.EventRequest,
    responseType: wechaty_puppet_event_pb.EventResponse,
    requestSerialize: serialize_wechaty_puppet_event_EventRequest,
    requestDeserialize: deserialize_wechaty_puppet_event_EventRequest,
    responseSerialize: serialize_wechaty_puppet_event_EventResponse,
    responseDeserialize: deserialize_wechaty_puppet_event_EventResponse,
  },
  // *
//
// Contact Self
//
contactSelfQRCode: {
    path: '/wechaty.puppet.Puppet/ContactSelfQRCode',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactSelfQRCodeRequest,
    responseType: wechaty_puppet_contact_pb.ContactSelfQRCodeResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactSelfQRCodeRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactSelfQRCodeRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactSelfQRCodeResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactSelfQRCodeResponse,
  },
  contactSelfName: {
    path: '/wechaty.puppet.Puppet/ContactSelfName',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactSelfNameRequest,
    responseType: wechaty_puppet_contact_pb.ContactSelfNameResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactSelfNameRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactSelfNameRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactSelfNameResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactSelfNameResponse,
  },
  contactSelfSignature: {
    path: '/wechaty.puppet.Puppet/ContactSelfSignature',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactSelfSignatureRequest,
    responseType: wechaty_puppet_contact_pb.ContactSelfSignatureResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactSelfSignatureRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactSelfSignatureRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactSelfSignatureResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactSelfSignatureResponse,
  },
  // *
//
// Contact
//
contactPayload: {
    path: '/wechaty.puppet.Puppet/ContactPayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactPayloadRequest,
    responseType: wechaty_puppet_contact_pb.ContactPayloadResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactPayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactPayloadRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactPayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactPayloadResponse,
  },
  contactAlias: {
    path: '/wechaty.puppet.Puppet/ContactAlias',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactAliasRequest,
    responseType: wechaty_puppet_contact_pb.ContactAliasResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactAliasRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactAliasRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactAliasResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactAliasResponse,
  },
  contactAvatar: {
    path: '/wechaty.puppet.Puppet/ContactAvatar',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactAvatarRequest,
    responseType: wechaty_puppet_contact_pb.ContactAvatarResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactAvatarRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactAvatarRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactAvatarResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactAvatarResponse,
  },
  contactList: {
    path: '/wechaty.puppet.Puppet/ContactList',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_contact_pb.ContactListRequest,
    responseType: wechaty_puppet_contact_pb.ContactListResponse,
    requestSerialize: serialize_wechaty_puppet_contact_ContactListRequest,
    requestDeserialize: deserialize_wechaty_puppet_contact_ContactListRequest,
    responseSerialize: serialize_wechaty_puppet_contact_ContactListResponse,
    responseDeserialize: deserialize_wechaty_puppet_contact_ContactListResponse,
  },
  // *
//
// Friendship
//
friendshipPayload: {
    path: '/wechaty.puppet.Puppet/FriendshipPayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_friendship_pb.FriendshipPayloadRequest,
    responseType: wechaty_puppet_friendship_pb.FriendshipPayloadResponse,
    requestSerialize: serialize_wechaty_puppet_friendship_FriendshipPayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_friendship_FriendshipPayloadRequest,
    responseSerialize: serialize_wechaty_puppet_friendship_FriendshipPayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_friendship_FriendshipPayloadResponse,
  },
  friendshipSearchPhone: {
    path: '/wechaty.puppet.Puppet/FriendshipSearchPhone',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_friendship_pb.FriendshipSearchPhoneRequest,
    responseType: wechaty_puppet_friendship_pb.FriendshipSearchPhoneResponse,
    requestSerialize: serialize_wechaty_puppet_friendship_FriendshipSearchPhoneRequest,
    requestDeserialize: deserialize_wechaty_puppet_friendship_FriendshipSearchPhoneRequest,
    responseSerialize: serialize_wechaty_puppet_friendship_FriendshipSearchPhoneResponse,
    responseDeserialize: deserialize_wechaty_puppet_friendship_FriendshipSearchPhoneResponse,
  },
  friendshipSearchWeixin: {
    path: '/wechaty.puppet.Puppet/FriendshipSearchWeixin',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_friendship_pb.FriendshipSearchWeixinRequest,
    responseType: wechaty_puppet_friendship_pb.FriendshipSearchWeixinResponse,
    requestSerialize: serialize_wechaty_puppet_friendship_FriendshipSearchWeixinRequest,
    requestDeserialize: deserialize_wechaty_puppet_friendship_FriendshipSearchWeixinRequest,
    responseSerialize: serialize_wechaty_puppet_friendship_FriendshipSearchWeixinResponse,
    responseDeserialize: deserialize_wechaty_puppet_friendship_FriendshipSearchWeixinResponse,
  },
  friendshipAdd: {
    path: '/wechaty.puppet.Puppet/FriendshipAdd',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_friendship_pb.FriendshipAddRequest,
    responseType: wechaty_puppet_friendship_pb.FriendshipAddResponse,
    requestSerialize: serialize_wechaty_puppet_friendship_FriendshipAddRequest,
    requestDeserialize: deserialize_wechaty_puppet_friendship_FriendshipAddRequest,
    responseSerialize: serialize_wechaty_puppet_friendship_FriendshipAddResponse,
    responseDeserialize: deserialize_wechaty_puppet_friendship_FriendshipAddResponse,
  },
  frendshipAccept: {
    path: '/wechaty.puppet.Puppet/FrendshipAccept',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_friendship_pb.FriendshipAcceptRequest,
    responseType: wechaty_puppet_friendship_pb.FriendshipAcceptResponse,
    requestSerialize: serialize_wechaty_puppet_friendship_FriendshipAcceptRequest,
    requestDeserialize: deserialize_wechaty_puppet_friendship_FriendshipAcceptRequest,
    responseSerialize: serialize_wechaty_puppet_friendship_FriendshipAcceptResponse,
    responseDeserialize: deserialize_wechaty_puppet_friendship_FriendshipAcceptResponse,
  },
  // *
//
// Message
//
messagePayload: {
    path: '/wechaty.puppet.Puppet/MessagePayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessagePayloadRequest,
    responseType: wechaty_puppet_message_pb.MessagePayloadResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessagePayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessagePayloadRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessagePayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessagePayloadResponse,
  },
  messageContact: {
    path: '/wechaty.puppet.Puppet/MessageContact',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageContactRequest,
    responseType: wechaty_puppet_message_pb.MessageContactResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageContactRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageContactRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageContactResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageContactResponse,
  },
  messageFile: {
    path: '/wechaty.puppet.Puppet/MessageFile',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageFileRequest,
    responseType: wechaty_puppet_message_pb.MessageFileResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageFileRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageFileRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageFileResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageFileResponse,
  },
  messageMiniProgram: {
    path: '/wechaty.puppet.Puppet/MessageMiniProgram',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageMiniProgramRequest,
    responseType: wechaty_puppet_message_pb.MessageMiniProgramResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageMiniProgramRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageMiniProgramRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageMiniProgramResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageMiniProgramResponse,
  },
  messageUrl: {
    path: '/wechaty.puppet.Puppet/MessageUrl',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageUrlRequest,
    responseType: wechaty_puppet_message_pb.MessageUrlResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageUrlRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageUrlRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageUrlResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageUrlResponse,
  },
  messageSendContact: {
    path: '/wechaty.puppet.Puppet/MessageSendContact',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageSendContactRequest,
    responseType: wechaty_puppet_message_pb.MessageSendContactResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageSendContactRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageSendContactRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageSendContactResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageSendContactResponse,
  },
  messageSendFile: {
    path: '/wechaty.puppet.Puppet/MessageSendFile',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageSendFileRequest,
    responseType: wechaty_puppet_message_pb.MessageSendFileResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageSendFileRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageSendFileRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageSendFileResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageSendFileResponse,
  },
  messageSendText: {
    path: '/wechaty.puppet.Puppet/MessageSendText',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageSendTextRequest,
    responseType: wechaty_puppet_message_pb.MessageSendTextResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageSendTextRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageSendTextRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageSendTextResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageSendTextResponse,
  },
  messageSendMiniProgram: {
    path: '/wechaty.puppet.Puppet/MessageSendMiniProgram',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageSendMiniProgramRequest,
    responseType: wechaty_puppet_message_pb.MessageSendMiniProgramResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageSendMiniProgramRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageSendMiniProgramRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageSendMiniProgramResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageSendMiniProgramResponse,
  },
  messageSendUrl: {
    path: '/wechaty.puppet.Puppet/MessageSendUrl',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageSendUrlRequest,
    responseType: wechaty_puppet_message_pb.MessageSendUrlResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageSendUrlRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageSendUrlRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageSendUrlResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageSendUrlResponse,
  },
  messageRecall: {
    path: '/wechaty.puppet.Puppet/MessageRecall',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_message_pb.MessageRecallRequest,
    responseType: wechaty_puppet_message_pb.MessageRecallResponse,
    requestSerialize: serialize_wechaty_puppet_message_MessageRecallRequest,
    requestDeserialize: deserialize_wechaty_puppet_message_MessageRecallRequest,
    responseSerialize: serialize_wechaty_puppet_message_MessageRecallResponse,
    responseDeserialize: deserialize_wechaty_puppet_message_MessageRecallResponse,
  },
  // *
//
// Room
//
roomPayload: {
    path: '/wechaty.puppet.Puppet/RoomPayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomPayloadRequest,
    responseType: wechaty_puppet_room_pb.RoomPayloadResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomPayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomPayloadRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomPayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomPayloadResponse,
  },
  roomList: {
    path: '/wechaty.puppet.Puppet/RoomList',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomListRequest,
    responseType: wechaty_puppet_room_pb.RoomListResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomListRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomListRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomListResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomListResponse,
  },
  roomAdd: {
    path: '/wechaty.puppet.Puppet/RoomAdd',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomAddRequest,
    responseType: wechaty_puppet_room_pb.RoomAddResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomAddRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomAddRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomAddResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomAddResponse,
  },
  roomAvatar: {
    path: '/wechaty.puppet.Puppet/RoomAvatar',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomAvatarRequest,
    responseType: wechaty_puppet_room_pb.RoomAvatarResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomAvatarRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomAvatarRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomAvatarResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomAvatarResponse,
  },
  roomCreate: {
    path: '/wechaty.puppet.Puppet/RoomCreate',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomCreateRequest,
    responseType: wechaty_puppet_room_pb.RoomCreateResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomCreateRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomCreateRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomCreateResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomCreateResponse,
  },
  roomDel: {
    path: '/wechaty.puppet.Puppet/RoomDel',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomDelRequest,
    responseType: wechaty_puppet_room_pb.RoomDelResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomDelRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomDelRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomDelResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomDelResponse,
  },
  roomQuit: {
    path: '/wechaty.puppet.Puppet/RoomQuit',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomQuitRequest,
    responseType: wechaty_puppet_room_pb.RoomQuitResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomQuitRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomQuitRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomQuitResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomQuitResponse,
  },
  roomTopic: {
    path: '/wechaty.puppet.Puppet/RoomTopic',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomTopicRequest,
    responseType: wechaty_puppet_room_pb.RoomTopicResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomTopicRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomTopicRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomTopicResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomTopicResponse,
  },
  roomQRCode: {
    path: '/wechaty.puppet.Puppet/RoomQRCode',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomQRCodeRequest,
    responseType: wechaty_puppet_room_pb.RoomQRCodeResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomQRCodeRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomQRCodeRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomQRCodeResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomQRCodeResponse,
  },
  roomAnnounce: {
    path: '/wechaty.puppet.Puppet/RoomAnnounce',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_pb.RoomAnnounceRequest,
    responseType: wechaty_puppet_room_pb.RoomAnnounceResponse,
    requestSerialize: serialize_wechaty_puppet_room_RoomAnnounceRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_RoomAnnounceRequest,
    responseSerialize: serialize_wechaty_puppet_room_RoomAnnounceResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_RoomAnnounceResponse,
  },
  // *
//
// Room Member
//
roomMemberPayload: {
    path: '/wechaty.puppet.Puppet/RoomMemberPayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_member_pb.RoomMemberPayloadRequest,
    responseType: wechaty_puppet_room_member_pb.RoomMemberPayloadResponse,
    requestSerialize: serialize_wechaty_puppet_room_member_RoomMemberPayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_member_RoomMemberPayloadRequest,
    responseSerialize: serialize_wechaty_puppet_room_member_RoomMemberPayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_member_RoomMemberPayloadResponse,
  },
  roomMemberList: {
    path: '/wechaty.puppet.Puppet/RoomMemberList',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_member_pb.RoomMemberListRequest,
    responseType: wechaty_puppet_room_member_pb.RoomMemberListResponse,
    requestSerialize: serialize_wechaty_puppet_room_member_RoomMemberListRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_member_RoomMemberListRequest,
    responseSerialize: serialize_wechaty_puppet_room_member_RoomMemberListResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_member_RoomMemberListResponse,
  },
  // *
//
// Room Invitation
//
roomInvitationPayload: {
    path: '/wechaty.puppet.Puppet/RoomInvitationPayload',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_invitation_pb.RoomInvitationPayloadRequest,
    responseType: wechaty_puppet_room_invitation_pb.RoomInvitationPayloadResponse,
    requestSerialize: serialize_wechaty_puppet_room_invitation_RoomInvitationPayloadRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_invitation_RoomInvitationPayloadRequest,
    responseSerialize: serialize_wechaty_puppet_room_invitation_RoomInvitationPayloadResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_invitation_RoomInvitationPayloadResponse,
  },
  roomInvitationAccept: {
    path: '/wechaty.puppet.Puppet/RoomInvitationAccept',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_room_invitation_pb.RoomInvitationAcceptRequest,
    responseType: wechaty_puppet_room_invitation_pb.RoomInvitationAcceptResponse,
    requestSerialize: serialize_wechaty_puppet_room_invitation_RoomInvitationAcceptRequest,
    requestDeserialize: deserialize_wechaty_puppet_room_invitation_RoomInvitationAcceptRequest,
    responseSerialize: serialize_wechaty_puppet_room_invitation_RoomInvitationAcceptResponse,
    responseDeserialize: deserialize_wechaty_puppet_room_invitation_RoomInvitationAcceptResponse,
  },
  // *
//
// Tag
//
tagContactAdd: {
    path: '/wechaty.puppet.Puppet/TagContactAdd',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_tag_pb.TagContactAddRequest,
    responseType: wechaty_puppet_tag_pb.TagContactAddResponse,
    requestSerialize: serialize_wechaty_puppet_tag_TagContactAddRequest,
    requestDeserialize: deserialize_wechaty_puppet_tag_TagContactAddRequest,
    responseSerialize: serialize_wechaty_puppet_tag_TagContactAddResponse,
    responseDeserialize: deserialize_wechaty_puppet_tag_TagContactAddResponse,
  },
  tagContactRemove: {
    path: '/wechaty.puppet.Puppet/TagContactRemove',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_tag_pb.TagContactRemoveRequest,
    responseType: wechaty_puppet_tag_pb.TagContactRemoveResponse,
    requestSerialize: serialize_wechaty_puppet_tag_TagContactRemoveRequest,
    requestDeserialize: deserialize_wechaty_puppet_tag_TagContactRemoveRequest,
    responseSerialize: serialize_wechaty_puppet_tag_TagContactRemoveResponse,
    responseDeserialize: deserialize_wechaty_puppet_tag_TagContactRemoveResponse,
  },
  tagContactDelete: {
    path: '/wechaty.puppet.Puppet/TagContactDelete',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_tag_pb.TagContactDeleteRequest,
    responseType: wechaty_puppet_tag_pb.TagContactDeleteResponse,
    requestSerialize: serialize_wechaty_puppet_tag_TagContactDeleteRequest,
    requestDeserialize: deserialize_wechaty_puppet_tag_TagContactDeleteRequest,
    responseSerialize: serialize_wechaty_puppet_tag_TagContactDeleteResponse,
    responseDeserialize: deserialize_wechaty_puppet_tag_TagContactDeleteResponse,
  },
  tagContactList: {
    path: '/wechaty.puppet.Puppet/TagContactList',
    requestStream: false,
    responseStream: false,
    requestType: wechaty_puppet_tag_pb.TagContactListRequest,
    responseType: wechaty_puppet_tag_pb.TagContactListResponse,
    requestSerialize: serialize_wechaty_puppet_tag_TagContactListRequest,
    requestDeserialize: deserialize_wechaty_puppet_tag_TagContactListRequest,
    responseSerialize: serialize_wechaty_puppet_tag_TagContactListResponse,
    responseDeserialize: deserialize_wechaty_puppet_tag_TagContactListResponse,
  },
};

exports.PuppetClient = grpc.makeGenericClientConstructor(PuppetService);
