// wechaty-puppet-hostie/dist/src/puppet-hostie-server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("./config");
var PuppetHostieServerType;
(function (PuppetHostieServerType) {
    PuppetHostieServerType[PuppetHostieServerType["Unknown"] = 0] = "Unknown";
    PuppetHostieServerType[PuppetHostieServerType["Grpc"] = 1] = "Grpc";
    PuppetHostieServerType[PuppetHostieServerType["JsonRpc"] = 2] = "JsonRpc";
    PuppetHostieServerType[PuppetHostieServerType["OpenApi"] = 3] = "OpenApi";
})(PuppetHostieServerType = exports.PuppetHostieServerType || (exports.PuppetHostieServerType = {}));
class PuppetHostieServer {
    constructor(options) {
        this.options = options;
        config_1.log.verbose('PuppetHostieServer', 'constructor(%s)', JSON.stringify(options));
    }
    version() {
        return config_1.VERSION;
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('PuppetHostieServer', `start()`);
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('PuppetHostieServer', `stop()`);
        });
    }
}
exports.PuppetHostieServer = PuppetHostieServer;
