// wechaty-puppet-padplus/dist/src/schemas/grpc-schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GRPC_CODE;
(function (GRPC_CODE) {
    GRPC_CODE[GRPC_CODE["OK"] = 0] = "OK";
    GRPC_CODE[GRPC_CODE["CANCELLED"] = 1] = "CANCELLED";
    GRPC_CODE[GRPC_CODE["UNKNOWN"] = 2] = "UNKNOWN";
    GRPC_CODE[GRPC_CODE["INVALID_ARGUMENT"] = 3] = "INVALID_ARGUMENT";
    GRPC_CODE[GRPC_CODE["DEADLINE_EXCEEDED"] = 4] = "DEADLINE_EXCEEDED";
    GRPC_CODE[GRPC_CODE["NOT_FOUND"] = 5] = "NOT_FOUND";
    GRPC_CODE[GRPC_CODE["ALREADY_EXISTS"] = 6] = "ALREADY_EXISTS";
    GRPC_CODE[GRPC_CODE["PERMISSION_DENIED"] = 7] = "PERMISSION_DENIED";
    GRPC_CODE[GRPC_CODE["UNAUTHENTICATED"] = 16] = "UNAUTHENTICATED";
    GRPC_CODE[GRPC_CODE["RESOURCE_EXHAUSTED"] = 8] = "RESOURCE_EXHAUSTED";
    GRPC_CODE[GRPC_CODE["FAILED_PRECONDITION"] = 9] = "FAILED_PRECONDITION";
    GRPC_CODE[GRPC_CODE["ABORTED"] = 10] = "ABORTED";
    GRPC_CODE[GRPC_CODE["OUT_OF_RANGE"] = 11] = "OUT_OF_RANGE";
    GRPC_CODE[GRPC_CODE["UNIMPLEMENTED"] = 12] = "UNIMPLEMENTED";
    GRPC_CODE[GRPC_CODE["INTERNAL"] = 13] = "INTERNAL";
    GRPC_CODE[GRPC_CODE["UNAVAILABLE"] = 14] = "UNAVAILABLE";
    GRPC_CODE[GRPC_CODE["DATA_LOSS"] = 15] = "DATA_LOSS";
})(GRPC_CODE = exports.GRPC_CODE || (exports.GRPC_CODE = {}));
