// wechaty-puppet-padplus/dist/src/schemas/padplus-enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var QrcodeStatus;
(function (QrcodeStatus) {
    QrcodeStatus[QrcodeStatus["Waiting"] = 0] = "Waiting";
    QrcodeStatus[QrcodeStatus["Scanned"] = 1] = "Scanned";
    QrcodeStatus[QrcodeStatus["Confirmed"] = 2] = "Confirmed";
    QrcodeStatus[QrcodeStatus["Canceled"] = 4] = "Canceled";
    QrcodeStatus[QrcodeStatus["Expired"] = 3] = "Expired";
})(QrcodeStatus = exports.QrcodeStatus || (exports.QrcodeStatus = {}));
var LoginStatus;
(function (LoginStatus) {
    LoginStatus[LoginStatus["Logined"] = 1] = "Logined";
})(LoginStatus = exports.LoginStatus || (exports.LoginStatus = {}));
var ContactType;
(function (ContactType) {
    ContactType[ContactType["Unknown"] = 0] = "Unknown";
    ContactType[ContactType["Personal"] = 99990] = "Personal";
    ContactType[ContactType["Official"] = 99991] = "Official";
})(ContactType = exports.ContactType || (exports.ContactType = {}));
var FriendshipType;
(function (FriendshipType) {
    FriendshipType[FriendshipType["Unknown"] = 0] = "Unknown";
    FriendshipType[FriendshipType["Confirm"] = 99990] = "Confirm";
    FriendshipType[FriendshipType["Receive"] = 99991] = "Receive";
    FriendshipType[FriendshipType["Verify"] = 99992] = "Verify";
})(FriendshipType = exports.FriendshipType || (exports.FriendshipType = {}));
var CheckQRCodeStatus;
(function (CheckQRCodeStatus) {
    CheckQRCodeStatus[CheckQRCodeStatus["Ignore"] = -2] = "Ignore";
    CheckQRCodeStatus[CheckQRCodeStatus["Unknown"] = -1] = "Unknown";
    CheckQRCodeStatus[CheckQRCodeStatus["WaitScan"] = 0] = "WaitScan";
    CheckQRCodeStatus[CheckQRCodeStatus["WaitConfirm"] = 1] = "WaitConfirm";
    CheckQRCodeStatus[CheckQRCodeStatus["Confirmed"] = 2] = "Confirmed";
    CheckQRCodeStatus[CheckQRCodeStatus["Timeout"] = 3] = "Timeout";
    CheckQRCodeStatus[CheckQRCodeStatus["Cancel"] = 4] = "Cancel";
})(CheckQRCodeStatus = exports.CheckQRCodeStatus || (exports.CheckQRCodeStatus = {}));
var RoomAddTypeStatus;
(function (RoomAddTypeStatus) {
    RoomAddTypeStatus[RoomAddTypeStatus["Done"] = 0] = "Done";
    RoomAddTypeStatus[RoomAddTypeStatus["NeedInvite"] = -2012] = "NeedInvite";
    RoomAddTypeStatus[RoomAddTypeStatus["InviteConfirm"] = -2028] = "InviteConfirm";
})(RoomAddTypeStatus = exports.RoomAddTypeStatus || (exports.RoomAddTypeStatus = {}));
/**
 * Raw type info:
 * see more inhttps://ymiao.oss-cn-shanghai.aliyuncs.com/apifile.txt
 * 2  - 通过搜索邮箱
 * 3  - 通过微信号搜索
 * 5  - 通过朋友验证消息
 * 7  - 通过朋友验证消息(可回复)
 * 12 - 通过QQ好友添加
 * 14 - 通过群来源
 * 15 - 通过搜索手机号
 * 16 - 通过朋友验证消息
 * 17 - 通过名片分享
 * 22 - 通过摇一摇打招呼方式
 * 25 - 通过漂流瓶
 * 30 - 通过二维码方式
 */
var SearchContactTypeStatus;
(function (SearchContactTypeStatus) {
    SearchContactTypeStatus[SearchContactTypeStatus["CONTACT"] = 17] = "CONTACT";
    SearchContactTypeStatus[SearchContactTypeStatus["EMAIL"] = 2] = "EMAIL";
    SearchContactTypeStatus[SearchContactTypeStatus["FLOAT"] = 25] = "FLOAT";
    SearchContactTypeStatus[SearchContactTypeStatus["MOBILE"] = 15] = "MOBILE";
    SearchContactTypeStatus[SearchContactTypeStatus["QQ"] = 12] = "QQ";
    SearchContactTypeStatus[SearchContactTypeStatus["QRCODE"] = 30] = "QRCODE";
    SearchContactTypeStatus[SearchContactTypeStatus["ROOM"] = 14] = "ROOM";
    SearchContactTypeStatus[SearchContactTypeStatus["Searchable"] = 0] = "Searchable";
    SearchContactTypeStatus[SearchContactTypeStatus["SHAKE"] = 22] = "SHAKE";
    SearchContactTypeStatus[SearchContactTypeStatus["UnSearchable"] = -24] = "UnSearchable";
    SearchContactTypeStatus[SearchContactTypeStatus["VERIFY"] = 16] = "VERIFY";
    SearchContactTypeStatus[SearchContactTypeStatus["VERIFY_NOREPLY"] = 5] = "VERIFY_NOREPLY";
    SearchContactTypeStatus[SearchContactTypeStatus["VERIFY_REPLY"] = 7] = "VERIFY_REPLY";
    SearchContactTypeStatus[SearchContactTypeStatus["WXID"] = 3] = "WXID";
})(SearchContactTypeStatus = exports.SearchContactTypeStatus || (exports.SearchContactTypeStatus = {}));
var PadplusMessageStatus;
(function (PadplusMessageStatus) {
    PadplusMessageStatus[PadplusMessageStatus["One"] = 1] = "One";
})(PadplusMessageStatus = exports.PadplusMessageStatus || (exports.PadplusMessageStatus = {}));
var PadplusStatus;
(function (PadplusStatus) {
    PadplusStatus[PadplusStatus["One"] = 1] = "One";
})(PadplusStatus = exports.PadplusStatus || (exports.PadplusStatus = {}));
var PadplusContinue;
(function (PadplusContinue) {
    PadplusContinue[PadplusContinue["Done"] = 0] = "Done";
    PadplusContinue[PadplusContinue["Go"] = 1] = "Go";
})(PadplusContinue = exports.PadplusContinue || (exports.PadplusContinue = {}));
var PadplusPayloadType;
(function (PadplusPayloadType) {
    PadplusPayloadType[PadplusPayloadType["ExpirePadplusToken"] = -1113] = "ExpirePadplusToken";
    PadplusPayloadType[PadplusPayloadType["InvalidPadplusToken"] = -1111] = "InvalidPadplusToken";
    PadplusPayloadType[PadplusPayloadType["Logout"] = -1] = "Logout";
    PadplusPayloadType[PadplusPayloadType["OnlinePadplusToken"] = -1112] = "OnlinePadplusToken";
})(PadplusPayloadType = exports.PadplusPayloadType || (exports.PadplusPayloadType = {}));
var WechatAppMessageType;
(function (WechatAppMessageType) {
    WechatAppMessageType[WechatAppMessageType["Text"] = 1] = "Text";
    WechatAppMessageType[WechatAppMessageType["Img"] = 2] = "Img";
    WechatAppMessageType[WechatAppMessageType["Audio"] = 3] = "Audio";
    WechatAppMessageType[WechatAppMessageType["Video"] = 4] = "Video";
    WechatAppMessageType[WechatAppMessageType["Url"] = 5] = "Url";
    WechatAppMessageType[WechatAppMessageType["Attach"] = 6] = "Attach";
    WechatAppMessageType[WechatAppMessageType["Open"] = 7] = "Open";
    WechatAppMessageType[WechatAppMessageType["Emoji"] = 8] = "Emoji";
    WechatAppMessageType[WechatAppMessageType["VoiceRemind"] = 9] = "VoiceRemind";
    WechatAppMessageType[WechatAppMessageType["ScanGood"] = 10] = "ScanGood";
    WechatAppMessageType[WechatAppMessageType["Good"] = 13] = "Good";
    WechatAppMessageType[WechatAppMessageType["Emotion"] = 15] = "Emotion";
    WechatAppMessageType[WechatAppMessageType["CardTicket"] = 16] = "CardTicket";
    WechatAppMessageType[WechatAppMessageType["RealtimeShareLocation"] = 17] = "RealtimeShareLocation";
    WechatAppMessageType[WechatAppMessageType["ChatHistory"] = 19] = "ChatHistory";
    WechatAppMessageType[WechatAppMessageType["MiniProgram"] = 33] = "MiniProgram";
    WechatAppMessageType[WechatAppMessageType["MiniProgramApp"] = 36] = "MiniProgramApp";
    WechatAppMessageType[WechatAppMessageType["Transfers"] = 2000] = "Transfers";
    WechatAppMessageType[WechatAppMessageType["RedEnvelopes"] = 2001] = "RedEnvelopes";
    WechatAppMessageType[WechatAppMessageType["ReaderType"] = 100001] = "ReaderType";
})(WechatAppMessageType = exports.WechatAppMessageType || (exports.WechatAppMessageType = {}));
var PadplusEmojiType;
(function (PadplusEmojiType) {
    PadplusEmojiType[PadplusEmojiType["Unknown"] = 0] = "Unknown";
    PadplusEmojiType[PadplusEmojiType["Static"] = 1] = "Static";
    PadplusEmojiType[PadplusEmojiType["Dynamic"] = 2] = "Dynamic";
})(PadplusEmojiType = exports.PadplusEmojiType || (exports.PadplusEmojiType = {}));
/**
 * Enum for MsgType values.
 * @enum {number}
 * @property {number} TEXT                - MsgType.TEXT                (1)     for TEXT
 * @property {number} IMAGE               - MsgType.IMAGE               (3)     for IMAGE
 * @property {number} VOICE               - MsgType.VOICE               (34)    for VOICE
 * @property {number} VERIFYMSG           - MsgType.VERIFYMSG           (37)    for VERIFYMSG
 * @property {number} POSSIBLEFRIEND_MSG  - MsgType.POSSIBLEFRIEND_MSG  (40)    for POSSIBLEFRIEND_MSG
 * @property {number} SHARECARD           - MsgType.SHARECARD           (42)    for SHARECARD
 * @property {number} VIDEO               - MsgType.VIDEO               (43)    for VIDEO
 * @property {number} EMOTICON            - MsgType.EMOTICON            (47)    for EMOTICON
 * @property {number} LOCATION            - MsgType.LOCATION            (48)    for LOCATION
 * @property {number} APP                 - MsgType.APP                 (49)    for APP         | File, Media Link
 * @property {number} VOIPMSG             - MsgType.VOIPMSG             (50)    for VOIPMSG
 * @property {number} STATUSNOTIFY        - MsgType.STATUSNOTIFY        (51)    for STATUSNOTIFY
 * @property {number} VOIPNOTIFY          - MsgType.VOIPNOTIFY          (52)    for VOIPNOTIFY
 * @property {number} VOIPINVITE          - MsgType.VOIPINVITE          (53)    for VOIPINVITE
 * @property {number} MICROVIDEO          - MsgType.MICROVIDEO          (62)    for MICROVIDEO
 * @property {number} SYSNOTICE           - MsgType.SYSNOTICE           (9999)  for SYSNOTICE
 * @property {number} SYS                 - MsgType.SYS                 (10000) for SYS         | Change Room Topic, Invite into Room, Kick Off from the room
 * @property {number} RECALLED            - MsgType.RECALLED            (10002) for RECALLED
 */
var PadplusMessageType;
(function (PadplusMessageType) {
    PadplusMessageType[PadplusMessageType["Text"] = 1] = "Text";
    PadplusMessageType[PadplusMessageType["Contact"] = 2] = "Contact";
    PadplusMessageType[PadplusMessageType["Image"] = 3] = "Image";
    PadplusMessageType[PadplusMessageType["Deleted"] = 4] = "Deleted";
    PadplusMessageType[PadplusMessageType["Voice"] = 34] = "Voice";
    PadplusMessageType[PadplusMessageType["SelfAvatar"] = 35] = "SelfAvatar";
    PadplusMessageType[PadplusMessageType["VerifyMsg"] = 37] = "VerifyMsg";
    PadplusMessageType[PadplusMessageType["PossibleFriendMsg"] = 40] = "PossibleFriendMsg";
    PadplusMessageType[PadplusMessageType["ShareCard"] = 42] = "ShareCard";
    PadplusMessageType[PadplusMessageType["Video"] = 43] = "Video";
    PadplusMessageType[PadplusMessageType["Emoticon"] = 47] = "Emoticon";
    PadplusMessageType[PadplusMessageType["Location"] = 48] = "Location";
    PadplusMessageType[PadplusMessageType["App"] = 49] = "App";
    PadplusMessageType[PadplusMessageType["VoipMsg"] = 50] = "VoipMsg";
    PadplusMessageType[PadplusMessageType["StatusNotify"] = 51] = "StatusNotify";
    PadplusMessageType[PadplusMessageType["VoipNotify"] = 52] = "VoipNotify";
    PadplusMessageType[PadplusMessageType["VoipInvite"] = 53] = "VoipInvite";
    PadplusMessageType[PadplusMessageType["MicroVideo"] = 62] = "MicroVideo";
    PadplusMessageType[PadplusMessageType["SelfInfo"] = 101] = "SelfInfo";
    PadplusMessageType[PadplusMessageType["SysNotice"] = 9999] = "SysNotice";
    PadplusMessageType[PadplusMessageType["Sys"] = 10000] = "Sys";
    PadplusMessageType[PadplusMessageType["Recalled"] = 10002] = "Recalled";
    PadplusMessageType[PadplusMessageType["N11_2048"] = 2048] = "N11_2048";
    PadplusMessageType[PadplusMessageType["N15_32768"] = 32768] = "N15_32768";
})(PadplusMessageType = exports.PadplusMessageType || (exports.PadplusMessageType = {}));
// TODO: figure out the meaning of the enum values
var PadplusRoomMemberFlag;
(function (PadplusRoomMemberFlag) {
    PadplusRoomMemberFlag[PadplusRoomMemberFlag["Zero"] = 0] = "Zero";
    PadplusRoomMemberFlag[PadplusRoomMemberFlag["One"] = 1] = "One";
    PadplusRoomMemberFlag[PadplusRoomMemberFlag["Eight"] = 8] = "Eight";
})(PadplusRoomMemberFlag = exports.PadplusRoomMemberFlag || (exports.PadplusRoomMemberFlag = {}));
var ContactOperationCmdId;
(function (ContactOperationCmdId) {
    ContactOperationCmdId[ContactOperationCmdId["Delete"] = 7] = "Delete";
    ContactOperationCmdId[ContactOperationCmdId["Operation"] = 2] = "Operation";
})(ContactOperationCmdId = exports.ContactOperationCmdId || (exports.ContactOperationCmdId = {}));
var ContactOperationBitVal;
(function (ContactOperationBitVal) {
    ContactOperationBitVal[ContactOperationBitVal["SaveToContact"] = 2051] = "SaveToContact";
    ContactOperationBitVal[ContactOperationBitVal["RemoveFromContact"] = 2] = "RemoveFromContact";
    ContactOperationBitVal[ContactOperationBitVal["Star"] = 71] = "Star";
    ContactOperationBitVal[ContactOperationBitVal["UnStar"] = 7] = "UnStar";
    ContactOperationBitVal[ContactOperationBitVal["Remark"] = 7] = "Remark";
    ContactOperationBitVal[ContactOperationBitVal["BlackList"] = 15] = "BlackList";
    ContactOperationBitVal[ContactOperationBitVal["UnBlackList"] = 7] = "UnBlackList";
})(ContactOperationBitVal = exports.ContactOperationBitVal || (exports.ContactOperationBitVal = {}));
var GrpcVoiceFormat;
(function (GrpcVoiceFormat) {
    GrpcVoiceFormat[GrpcVoiceFormat["Amr"] = 1] = "Amr";
    GrpcVoiceFormat[GrpcVoiceFormat["Mp3"] = 2] = "Mp3";
    GrpcVoiceFormat[GrpcVoiceFormat["Wave"] = 3] = "Wave";
    GrpcVoiceFormat[GrpcVoiceFormat["Silk"] = 4] = "Silk";
})(GrpcVoiceFormat = exports.GrpcVoiceFormat || (exports.GrpcVoiceFormat = {}));
var GrpcA8KeyScene;
(function (GrpcA8KeyScene) {
    GrpcA8KeyScene[GrpcA8KeyScene["ContactOrRoom"] = 2] = "ContactOrRoom";
    GrpcA8KeyScene[GrpcA8KeyScene["HistoryReading"] = 3] = "HistoryReading";
    GrpcA8KeyScene[GrpcA8KeyScene["QRCodeLink"] = 4] = "QRCodeLink";
    GrpcA8KeyScene[GrpcA8KeyScene["OAAccount"] = 7] = "OAAccount";
})(GrpcA8KeyScene = exports.GrpcA8KeyScene || (exports.GrpcA8KeyScene = {}));
var AutoLoginError;
(function (AutoLoginError) {
    AutoLoginError["CALL_FAILED"] = "CALL_FAILED";
    AutoLoginError["LOGIN_ERROR"] = "LOGIN_ERROR";
})(AutoLoginError = exports.AutoLoginError || (exports.AutoLoginError = {}));
var EncryptionServiceError;
(function (EncryptionServiceError) {
    EncryptionServiceError["NO_SESSION"] = "NO_SESSION";
    EncryptionServiceError["INTERNAL_ERROR"] = "INTERNAL_ERROR";
})(EncryptionServiceError = exports.EncryptionServiceError || (exports.EncryptionServiceError = {}));
var GrpcSelfAvatarType;
(function (GrpcSelfAvatarType) {
    GrpcSelfAvatarType[GrpcSelfAvatarType["CURRENT"] = 1] = "CURRENT";
    GrpcSelfAvatarType[GrpcSelfAvatarType["OLD"] = 2] = "OLD";
})(GrpcSelfAvatarType = exports.GrpcSelfAvatarType || (exports.GrpcSelfAvatarType = {}));
var CDNFileType;
(function (CDNFileType) {
    CDNFileType[CDNFileType["IMAGE"] = 1] = "IMAGE";
    CDNFileType[CDNFileType["MID_IMAGE"] = 2] = "MID_IMAGE";
    CDNFileType[CDNFileType["VIDEO_THUMBNAIL"] = 3] = "VIDEO_THUMBNAIL";
    CDNFileType[CDNFileType["VIDEO"] = 4] = "VIDEO";
    CDNFileType[CDNFileType["ATTACHMENT"] = 5] = "ATTACHMENT";
})(CDNFileType = exports.CDNFileType || (exports.CDNFileType = {}));
var CDNFileMd5Exist;
(function (CDNFileMd5Exist) {
    CDNFileMd5Exist[CDNFileMd5Exist["NON_EXIST"] = 0] = "NON_EXIST";
    CDNFileMd5Exist[CDNFileMd5Exist["EXIST"] = 1] = "EXIST";
})(CDNFileMd5Exist = exports.CDNFileMd5Exist || (exports.CDNFileMd5Exist = {}));
var PadplusErrorType;
(function (PadplusErrorType) {
    PadplusErrorType["LOGIN"] = "LOGIN";
    PadplusErrorType["NO_ID"] = "NO_ID";
    PadplusErrorType["NO_CACHE"] = "NO_CACHE";
    PadplusErrorType["EXIT"] = "EXIT";
})(PadplusErrorType = exports.PadplusErrorType || (exports.PadplusErrorType = {}));
var PadplusAutoLoginErrorType;
(function (PadplusAutoLoginErrorType) {
    PadplusAutoLoginErrorType["SELF_LOGOUT"] = "SELF_LOGOUT";
    PadplusAutoLoginErrorType["TOO_FREQUENT"] = "TOO_FREQUENT";
    PadplusAutoLoginErrorType["LOGIN_ANOTHER_DEVICE"] = "LOGIN_ANOTHER_DEVICE";
    PadplusAutoLoginErrorType["LOGIN_ANOTHER_DEVICE_WITH_WARN"] = "LOGIN_ANOTHER_DEVICE_WITH_WARN";
    PadplusAutoLoginErrorType["SAFETY_LOGOUT"] = "SAFETY_LOGOUT";
    PadplusAutoLoginErrorType["UNKNOWN"] = "UNKNOWN";
})(PadplusAutoLoginErrorType = exports.PadplusAutoLoginErrorType || (exports.PadplusAutoLoginErrorType = {}));
var MessageSendType;
(function (MessageSendType) {
    MessageSendType[MessageSendType["SELF_SENT"] = 1] = "SELF_SENT";
    MessageSendType[MessageSendType["CONTACT_SENT"] = 2] = "CONTACT_SENT";
})(MessageSendType = exports.MessageSendType || (exports.MessageSendType = {}));
var RequestStatus;
(function (RequestStatus) {
    RequestStatus[RequestStatus["Fail"] = 0] = "Fail";
    RequestStatus[RequestStatus["Success"] = 1] = "Success";
})(RequestStatus = exports.RequestStatus || (exports.RequestStatus = {}));
var GrpcMQType;
(function (GrpcMQType) {
    GrpcMQType[GrpcMQType["RECEIVE_MESSAGE"] = 2] = "RECEIVE_MESSAGE";
    GrpcMQType[GrpcMQType["CONTACT_INFO_CHANGE"] = 3] = "CONTACT_INFO_CHANGE";
    GrpcMQType[GrpcMQType["DELETE_CONTACT"] = 4] = "DELETE_CONTACT";
    GrpcMQType[GrpcMQType["GET_ROOM_MEMBER"] = 5] = "GET_ROOM_MEMBER";
    GrpcMQType[GrpcMQType["ROOM_MEMBER_CHANGE"] = 6] = "ROOM_MEMBER_CHANGE";
    GrpcMQType[GrpcMQType["GET_CONTACT"] = 7] = "GET_CONTACT";
    GrpcMQType[GrpcMQType["ADD_CONTACT"] = 15] = "ADD_CONTACT";
    GrpcMQType[GrpcMQType["SYNC"] = 51] = "SYNC";
    GrpcMQType[GrpcMQType["LOGOUT"] = 1100] = "LOGOUT";
})(GrpcMQType = exports.GrpcMQType || (exports.GrpcMQType = {}));
