// wechaty/dist/src/io-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 *   Wechaty - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016-2018 Huan LI <zixia@zixia.net>
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
/**
 * DO NOT use `require('../')` here!
 * because it will casue a LOOP require ERROR
 */
const state_switch_1 = require("state-switch");
const wechaty_puppet_hostie_1 = require("wechaty-puppet-hostie");
const config_1 = require("./config");
const io_1 = require("./io");
class IoClient {
    constructor(options) {
        this.options = options;
        config_1.log.verbose('IoClient', 'constructor({token: %s})', options.token);
        this.state = new state_switch_1.StateSwitch('IoClient', config_1.log);
    }
    startHostie() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'startHostie()');
            if (this.hostieServer) {
                throw new Error('hostie server exists');
            }
            const options = {
                endpoint: '0.0.0.0:8788',
                puppet: this.options.wechaty.puppet,
                token: this.options.token,
            };
            this.hostieServer = new wechaty_puppet_hostie_1.PuppetHostieServer(options);
            yield this.hostieServer.start();
        });
    }
    stopHostie() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'stopHostie()');
            if (!this.hostieServer) {
                throw new Error('hostie server does not exist');
            }
            yield this.hostieServer.stop();
            this.hostieServer = undefined;
        });
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'start()');
            if (this.state.pending()) {
                config_1.log.warn('IoClient', 'start() with a pending state, not the time');
                const e = new Error('state.pending() when start()');
                throw e;
            }
            this.state.on('pending');
            try {
                yield this.hookWechaty(this.options.wechaty);
                yield this.startIo();
                yield this.options.wechaty.start();
                yield this.startHostie();
                this.state.on(true);
            }
            catch (e) {
                config_1.log.error('IoClient', 'start() exception: %s', e.message);
                this.state.off(true);
                throw e;
            }
        });
    }
    hookWechaty(wechaty) {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'initWechaty()');
            if (this.state.off()) {
                const e = new Error('state.off() is true, skipped');
                config_1.log.warn('IoClient', 'initWechaty() %s', e.message);
                throw e;
            }
            wechaty
                .on('login', user => config_1.log.info('IoClient', `${user.name()} logined`))
                .on('logout', user => config_1.log.info('IoClient', `${user.name()} logouted`))
                .on('scan', (url, code) => config_1.log.info('IoClient', `[${code}] ${url}`))
                .on('message', msg => this.onMessage(msg));
        });
    }
    startIo() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'startIo() with token %s', this.options.token);
            if (this.state.off()) {
                const e = new Error('startIo() state.off() is true, skipped');
                config_1.log.warn('IoClient', e.message);
                throw e;
            }
            if (this.io) {
                throw new Error('io exists');
            }
            this.io = new io_1.Io({
                token: this.options.token,
                wechaty: this.options.wechaty,
            });
            try {
                yield this.io.start();
            }
            catch (e) {
                config_1.log.verbose('IoClient', 'startIo() init fail: %s', e.message);
                throw e;
            }
        });
    }
    stopIo() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'stopIo()');
            if (!this.io) {
                config_1.log.warn('IoClient', 'stopIo() io does not exist');
                return;
            }
            yield this.io.stop();
            this.io = undefined;
        });
    }
    onMessage(msg) {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'onMessage(%s)', msg);
            // const from = m.from()
            // const to = m.to()
            // const content = m.toString()
            // const room = m.room()
            // log.info('Bot', '%s<%s>:%s'
            //               , (room ? '['+room.topic()+']' : '')
            //               , from.name()
            //               , m.toStringDigest()
            //         )
            // if (/^wechaty|chatie|botie/i.test(m.text()) && !m.self()) {
            //   await m.say('https://www.chatie.io')
            //     .then(_ => log.info('Bot', 'REPLIED to magic word "chatie"'))
            // }
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'stop()');
            this.state.off('pending');
            yield this.stopIo();
            yield this.stopHostie();
            yield this.options.wechaty.stop();
            this.state.off(true);
            // XXX 20161026
            // this.io = null
        });
    }
    restart() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'restart()');
            try {
                yield this.stop();
                yield this.start();
            }
            catch (e) {
                config_1.log.error('IoClient', 'restart() exception %s', e.message);
                throw e;
            }
        });
    }
    quit() {
        return __awaiter(this, void 0, void 0, function* () {
            config_1.log.verbose('IoClient', 'quit()');
            if (this.state.off() === 'pending') {
                config_1.log.warn('IoClient', 'quit() with state.off() = `pending`, skipped');
                throw new Error('quit() with state.off() = `pending`');
            }
            this.state.off('pending');
            try {
                if (this.options.wechaty) {
                    yield this.options.wechaty.stop();
                    // this.wechaty = null
                }
                else {
                    config_1.log.warn('IoClient', 'quit() no this.wechaty');
                }
                if (this.io) {
                    yield this.io.stop();
                    // this.io = null
                }
                else {
                    config_1.log.warn('IoClient', 'quit() no this.io');
                }
            }
            catch (e) {
                config_1.log.error('IoClient', 'exception: %s', e.message);
                throw e;
            }
            finally {
                this.state.off(true);
            }
        });
    }
}
exports.IoClient = IoClient;
