// wechaty-puppet-padplus/dist/src/pure-function-helpers/room-member-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function briefRoomMemberParser(input) {
    const result = {};
    input.forEach(brief => {
        const memberPayload = {
            bigHeadUrl: '',
            contactId: brief.UserName,
            displayName: '',
            inviterId: '',
            nickName: brief.NickName || '',
            smallHeadUrl: '',
        };
        result[brief.UserName] = memberPayload;
    });
    return result;
}
exports.briefRoomMemberParser = briefRoomMemberParser;
function roomMemberParser(input) {
    const result = {};
    input.forEach(member => {
        const memberPayload = {
            bigHeadUrl: member.HeadImgUrl,
            contactId: member.UserName,
            displayName: member.DisplayName,
            inviterId: '',
            nickName: member.NickName,
            smallHeadUrl: member.HeadImgUrl,
        };
        result[member.UserName] = memberPayload;
    });
    return result;
}
exports.roomMemberParser = roomMemberParser;
