// wechaty-puppet/dist/src/schemas/puppet.js
"use strict";
// tslint:disable:object-literal-key-quotes
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This is used internally to as a placeholder for the bot name.
 *
 * For example:
 *  we should replace '你' and 'You' to YOU.
 *
 * See: https://github.com/Microsoft/TypeScript/issues/20898#issuecomment-354073352
 */
exports.YOU = Symbol('You');
/** @hidden */
exports.CHAT_EVENT_DICT = {
    friendship: 'receive a friend request',
    login: 'puppet had logined',
    logout: 'puppet had logouted',
    message: 'received a new message',
    'room-invite': 'received a room invitation',
    'room-join': 'be added to a room',
    'room-leave': 'leave or be removed from a room',
    'room-topic': 'room topic had been changed',
    scan: 'a QR Code scan is required',
};
/** @hidden */
exports.PUPPET_EVENT_DICT = Object.assign(Object.assign({}, exports.CHAT_EVENT_DICT), { dong: 'emit this event if you received a ding() call', error: `emit an Error instance when there's any Error need to report to Wechaty`, ready: 'emit this event after the puppet is ready(you define it)', reset: 'reset the puppet by emit this event', watchdog: 'feed the watchdog by emit this event' });
/**
 * @deprecated Huan(202002) use (conversationalId: string) instead.
 */
// export interface Receiver {
//   contactId? : string,
//   roomId?    : string,
// }
