// wechaty-puppet/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var contact_1 = require("./schemas/contact");
exports.ContactGender = contact_1.ContactGender;
exports.ContactType = contact_1.ContactType;
var event_1 = require("./schemas/event");
exports.ScanStatus = event_1.ScanStatus;
var friendship_1 = require("./schemas/friendship");
exports.FriendshipType = friendship_1.FriendshipType;
exports.FriendshipSceneType = friendship_1.FriendshipSceneType;
var image_1 = require("./schemas/image");
exports.ImageType = image_1.ImageType;
var message_1 = require("./schemas/message");
exports.MessageType = message_1.MessageType;
var throw_unsupported_error_1 = require("./throw-unsupported-error");
exports.throwUnsupportedError = throw_unsupported_error_1.throwUnsupportedError;
var puppet_1 = require("./schemas/puppet");
// Receiver,
exports.CHAT_EVENT_DICT = puppet_1.CHAT_EVENT_DICT;
exports.PUPPET_EVENT_DICT = puppet_1.PUPPET_EVENT_DICT;
exports.YOU = puppet_1.YOU;
var config_1 = require("./config");
exports.VERSION = config_1.VERSION;
exports.log = config_1.log;
var puppet_2 = require("./puppet");
exports.Puppet = puppet_2.Puppet;
